### PR TITLE
feat: cowswap approval needed approve infinite

### PR DESCRIPTION
**Pull request by chxash**
This implements `approvalNeeded` and `approveInfinite` in `CowSwapper`.
Due to the similarities with what is being done for Zrx regarding approval, I factorized some code in upper directory in swapper package, namely `getAllowanceRequired` and `grantAllowance` methods.

closes https://github.com/shapeshift/lib/issues/728

I was not able to test this in isolation in lib, nor integrate it into web by linking libs locally. I think I will need more methods implemented, and latest libs change to be supported in web, in order to effectively test this ... any suggestion for this is welcome also.

Testing : verify that related unit tests are meaningful and green

### CODE REVIEW

File: packages/swapper/src/swappers/utils/helpers/helpers.test.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
+import { HDWallet } from '@shapeshiftoss/hdwallet-core'
```
Comment (chxash) [inline packages/swapper/src/swappers/utils/helpers/helpers.test.ts:1]: I moved those tests along with `helpers.ts` from zrx to upper directory

**Comment by DaoDev44** [comment]
Awesome! Thanks for moving those things out of Zrx. I think were on the same page, because I have very similar changes in the [lib PR](https://github.com/shapeshift/lib/pull/765) I have up. Based on you changes, it *shouldn't* rug you too much. We have the same directory naming.

File: packages/swapper/src/swappers/cow/CowApprovalNeeded/CowApprovalNeeded.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    const adapter = await adapterManager.byChainId(sellAsset.chainId)
+    const receiveAddress = await adapter.getAddress({ wallet })
+
+    const allowanceResult = await getERC20Allowance({
+      web3,
+      erc20AllowanceAbi,
+      sellAssetErc20Address,
+      spenderAddress: COW_SWAP_VAULT_RELAYER_ADDRESS,
```
Comment (DaoDev44) [inline packages/swapper/src/swappers/cow/CowApprovalNeeded/CowApprovalNeeded.ts:35]: just curious, is there an api endpoint we can get this from?

Comment (chxash) [reply packages/swapper/src/swappers/cow/CowApprovalNeeded/CowApprovalNeeded.ts:35]: I looked but it doesn't seem to exist ...

**Review by DaoDev44 — DISMISSED** [review]

**Comment by chxash** [comment]
Did some painful merge of main into this

**Comment by elmutt** [comment]
> Did some painful merge of main into this

Is zrx is thoroughly tested after these changes?

Please test swaps from eth -> erc20, erc20 -> eth, and swaps requiring an approval first

**Review by 0xdef1cafe — APPROVED** [review]

**Comment by chxash** [comment]
Seen with @elmutt it seems pretty hard to link lib packages with web currently, so I am not able to verify Zrx thoroughly right now.

**Pull request merged**